File: Algorithms/710.Random-Pick-with-Blacklist/random-pick-with-blacklist.py
# -*- coding: utf-8 -*-
import random


class Solution:

    def __init__(self, N, blacklist):
        """
        :type N: int
        :type blacklist: List[int]
        """
        self.length = N - 1
        self.dic = {}
        blacklist.sort(reverse=True)
        for ele in blacklist:
            self.dic[ele] = self.dic.get(self.length, self.length)
            self.length -= 1
        

    def pick(self):
        """
        :rtype: int
        """
        num = random.randint(0, self.length)
        return self.dic.get(num, num)
        


# Your Solution object will be instantiated and called as such:
# obj = Solution(N, blacklist)
# param_1 = obj.pick()

File: Algorithms/710.Random-Pick-with-Blacklist/random_pick_with_blacklist.go
import (
	"math/rand"
	"sort"
)

type Solution struct {
	blackDic map[int]int
	length   int
}

func Constructor(N int, blacklist []int) Solution {
	blackDic := make(map[int]int, len(blacklist))
	length := N - 1

	sort.Sort(sort.Reverse(sort.IntSlice(blacklist)))

	for _, ele := range blacklist {
		t, ok := blackDic[length]
		if ok {
			blackDic[ele] = t
		} else {
			blackDic[ele] = length
		}
		length--
	}
	return Solution{
		length:   length,
		blackDic: blackDic,
	}

}

func (this *Solution) Pick() int {
	num := rand.Intn(this.length + 1)
	if t, ok := this.blackDic[num]; ok {
		return t
	}
	return num
}
